kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/interceptors/RetryInterceptor.kt

```diff
@@ -21,37 +26,58 @@ class RetryInterceptor(
     private val failureThreshold: Int = 3,
     private val invocationTimeout: Long = 0,
     private val ratio: Int = 1,
-    private val errorStatuses: List<Int> = listOf(503, 504)
+    private val errorStatuses: List<Int> = listOf(429, 503, 504)
 ) : Interceptor {
 
+    //https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Date#Syntax
+    private val httpDateFormat = SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss zzz")
+
     override fun intercept(chain: Interceptor.Chain): Response {
         val request = chain.request()
+        val maxDelayTime = chain.readTimeoutMillis()
         var attemptsCount = 0
         var delay = invocationTimeout
         while (true) {
             try {
-                if (shouldDelay(attemptsCount)) {
-                    delay = performAndReturnDelay(delay)
-                }
+                if (shouldDelay(attemptsCount)) performDelay(delay)
                 val response = chain.proceed(request)
-                if (!isRetry(response, attemptsCount)) return response
+                delay = calculateNextRetry(response.headers(), response.code(), attemptsCount, delay, maxDelayTime)
+                        ?: return response
             } catch (e: SocketTimeoutException) {
                 if (attemptsCount >= failureThreshold) throw e
             }
             attemptsCount++
         }
     }
 
-    internal fun performAndReturnDelay(delay: Long): Long {
+    private fun shouldDelay(attemptsCount: Int) = invocationTimeout > 0 && attemptsCount > 0
+
+    private fun performDelay(delay: Long) {
         try {
             Thread.sleep(delay)
         } catch (ignored: InterruptedException) {
         }
-        return delay * ratio
     }
 
-    private fun shouldDelay(attemptsCount: Int) = invocationTimeout > 0 && attemptsCount > 0
+    internal fun calculateNextRetry(responseHeaders: Headers, responseCode: Int, attemptsCount: Int, delay: Long, maxDelayTime: Int): Long? {
+        val retryAfter = parseRetryAfter(responseHeaders)
+        if (retryAfter != null && retryAfter > maxDelayTime) return null
 
-    internal fun isRetry(response: Response, attemptsCount: Int): Boolean =
-            attemptsCount < failureThreshold && response.code() in errorStatuses
+        return if (attemptsCount < failureThreshold && responseCode in errorStatuses) {
+            max(retryAfter ?: 0, delay * ratio)
+        } else {
+            null
+        }
+    }
+
+    internal fun parseRetryAfter(headers: Headers): Long? {
+        val retryAfter = headers["Retry-After"] ?: return null
+        return retryAfter.toLongOrNull()?.let { it * 1000 } ?: try {
+            val date = httpDateFormat.parse(retryAfter).toInstant().epochSecond
```

Comment:
Working with time is tricky. I strongly suggest several things:
1. Do not do any custom calculations. with timestamps(epoch, etc )
2. Use joda time primitives. It will take care of all things like Arabic, Japanese, or Thai calendars.   

        val responseDate = ZonedDateTime.parse("...")
        val now = ZonedDateTime.now()
        val duration = Duration.between(responseDate, now)
        
3. Please separate two stratagies in your code: 1st is to parse as number; 2nd is to parse as timestamp